<a href="https://colab.research.google.com/github/AnastaSya73/lalala/blob/main/%D0%9B%D0%B0%D0%B11.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Задание по практике 1

##Загрузка датасета
В файле classification datasets расположены ссылки на датасеты. Выберите один в соответствии со своим вариантом, скачайте и прочитайте файл
1

**Не забудьте загрузить нужные библиотеки**

Например: pip3 install pandas

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
df = pd.read_csv("/content/student-por.csv")#ваш код
pd.set_option('display.max_columns', None)
df

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   at_home   teacher   
1       GP   F   17       U     GT3       T     1     1   at_home     other   
2       GP   F   15       U     LE3       T     1     1   at_home     other   
3       GP   F   15       U     GT3       T     4     2    health  services   
4       GP   F   16       U     GT3       T     3     3     other     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
644     MS   F   19       R     GT3       T     2     3  services     other   
645     MS   F   18       U     LE3       T     3     1   teacher  services   
646     MS   F   18       U     GT3       T     1     1     other     other   
647     MS   M   17       U     LE3       T     3     1  services  services   
648     MS   M   18       R     LE3       T     3     2  services     other   

     reason guardian  traveltime  studytime  failures schoolsup famsup paid  \
0    course   mother           2          2         0       yes     no   no   
1    course   father           1          2         0        no    yes   no   
2     other   mother           1          2         0       yes     no   no   
3      home   mother           1          3         0        no    yes   no   
4      home   father           1          2         0        no    yes   no   
..      ...      ...         ...        ...       ...       ...    ...  ...   
644  course   mother           1          3         1        no     no   no   
645  course   mother           1          2         0        no    yes   no   
646  course   mother           2          2         0        no     no   no   
647  course   mother           2          1         0        no     no   no   
648  course   mother           3          1         0        no     no   no   

    activities nursery higher internet romantic  famrel  freetime  goout  \
0           no     yes    yes       no       no       4         3      4   
1           no      no    yes      yes       no       5         3      3   
2           no     yes    yes      yes       no       4         3      2   
3          yes     yes    yes      yes      yes       3         2      2   
4           no     yes    yes       no       no       4         3      2   
..         ...     ...    ...      ...      ...     ...       ...    ...   
644        yes      no    yes      yes       no       5         4      2   
645         no     yes    yes      yes       no       4         3      4   
646        yes     yes    yes       no       no       1         1      1   
647         no      no    yes      yes       no       2         4      5   
648         no      no    yes      yes       no       4         4      1   

     Dalc  Walc  health  absences  G1  G2  G3  
0       1     1       3         4   0  11  11  
1       1     1       3         2   9  11  11  
2       2     3       3         6  12  13  12  
3       1     1       5         0  14  14  14  
4       1     2       5         0  11  13  13  
..    ...   ...     ...       ...  ..  ..  ..  
644     1     2       5         4  10  11  10  
645     1     1       1         4  15  15  16  
646     1     1       5         6  11  12   9  
647     3     4       2         6  10  10  10  
648     3     4       5         4  10  11  11  

[649 rows x 33 columns]

##Первое знакомство с датасетом
В следующих ячейках выведите:


1.   Колонки датасета
2.   Первые 5 значений
3.   Последние 5 значений
4.   Статистику по числовым колонкам



In [ ]:
df = pd.read_csv("/content/student-por.csv")#Колонки

df.columns

Index(['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu',
       'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime',
       'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery',
       'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc',
       'Walc', 'health', 'absences', 'G1', 'G2', 'G3'],
      dtype='object')

In [ ]:
df[:5]#Первые 5 значений

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

   reason guardian  traveltime  studytime  failures schoolsup famsup paid  \
0  course   mother           2          2         0       yes     no   no   
1  course   father           1          2         0        no    yes   no   
2   other   mother           1          2         0       yes     no   no   
3    home   mother           1          3         0        no    yes   no   
4    home   father           1          2         0        no    yes   no   

  activities nursery higher internet romantic  famrel  freetime  goout  Dalc  \
0         no     yes    yes       no       no       4         3      4     1   
1         no      no    yes      yes       no       5         3      3     1   
2         no     yes    yes      yes       no       4         3      2     2   
3        yes     yes    yes      yes      yes       3         2      2     1   
4         no     yes    yes       no       no       4         3      2     1   

   Walc  health  absences  G1  G2  G3  
0     1       3         4   0  11  11  
1     1       3         2   9  11  11  
2     3       3         6  12  13  12  
3     1       5         0  14  14  14  
4     2       5         0  11  13  13

In [ ]:
df.tail()#Последние 5 значений

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
644     MS   F   19       R     GT3       T     2     3  services     other   
645     MS   F   18       U     LE3       T     3     1   teacher  services   
646     MS   F   18       U     GT3       T     1     1     other     other   
647     MS   M   17       U     LE3       T     3     1  services  services   
648     MS   M   18       R     LE3       T     3     2  services     other   

     reason guardian  traveltime  studytime  failures schoolsup famsup paid  \
644  course   mother           1          3         1        no     no   no   
645  course   mother           1          2         0        no    yes   no   
646  course   mother           2          2         0        no     no   no   
647  course   mother           2          1         0        no     no   no   
648  course   mother           3          1         0        no     no   no   

    activities nursery higher internet romantic  famrel  freetime  goout  \
644        yes      no    yes      yes       no       5         4      2   
645         no     yes    yes      yes       no       4         3      4   
646        yes     yes    yes       no       no       1         1      1   
647         no      no    yes      yes       no       2         4      5   
648         no      no    yes      yes       no       4         4      1   

     Dalc  Walc  health  absences  G1  G2  G3  
644     1     2       5         4  10  11  10  
645     1     1       1         4  15  15  16  
646     1     1       5         6  11  12   9  
647     3     4       2         6  10  10  10  
648     3     4       5         4  10  11  11

In [ ]:
df.describe()#Статистику

age        Medu        Fedu  traveltime   studytime    failures  \
count  649.000000  649.000000  649.000000  649.000000  649.000000  649.000000   
mean    16.744222    2.514638    2.306626    1.568567    1.930663    0.221880   
std      1.218138    1.134552    1.099931    0.748660    0.829510    0.593235   
min     15.000000    0.000000    0.000000    1.000000    1.000000    0.000000   
25%     16.000000    2.000000    1.000000    1.000000    1.000000    0.000000   
50%     17.000000    2.000000    2.000000    1.000000    2.000000    0.000000   
75%     18.000000    4.000000    3.000000    2.000000    2.000000    0.000000   
max     22.000000    4.000000    4.000000    4.000000    4.000000    3.000000   

           famrel    freetime       goout        Dalc        Walc      health  \
count  649.000000  649.000000  649.000000  649.000000  649.000000  649.000000   
mean     3.930663    3.180277    3.184900    1.502311    2.280431    3.536210   
std      0.955717    1.051093    1.175766    0.924834    1.284380    1.446259   
min      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   
25%      4.000000    3.000000    2.000000    1.000000    1.000000    2.000000   
50%      4.000000    3.000000    3.000000    1.000000    2.000000    4.000000   
75%      5.000000    4.000000    4.000000    2.000000    3.000000    5.000000   
max      5.000000    5.000000    5.000000    5.000000    5.000000    5.000000   

         absences          G1          G2          G3  
count  649.000000  649.000000  649.000000  649.000000  
mean     3.659476   11.399076   11.570108   11.906009  
std      4.640759    2.745265    2.913639    3.230656  
min      0.000000    0.000000    0.000000    0.000000  
25%      0.000000   10.000000   10.000000   10.000000  
50%      2.000000   11.000000   11.000000   12.000000  
75%      6.000000   13.000000   13.000000   14.000000  
max     32.000000   19.000000   19.000000   19.000000

**Ответьте на следующие вопросы:**


1.   Какие средние значения в каждой числовой колонке вашего датасета?
2.   Минимальные и максимальные значения в каждой числовой колонке вашего датасета



In [ ]:
df.mean()# средние значения в каждой числовой колонке датасета

<ipython-input-8-4ab5cb95ab61>:1: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df.mean()# средние значения в каждой числовой колонке датасета


age           16.744222
Medu           2.514638
Fedu           2.306626
traveltime     1.568567
studytime      1.930663
failures       0.221880
famrel         3.930663
freetime       3.180277
goout          3.184900
Dalc           1.502311
Walc           2.280431
health         3.536210
absences       3.659476
G1            11.399076
G2            11.570108
G3            11.906009
dtype: float64

In [ ]:
df_numerical = df.select_dtypes(include = [np.number])#выбираем из датасета столбцы с числовыми значениями
df_numerical.min()#Минимальные значения в каждой числовой колонке датасета

age           15
Medu           0
Fedu           0
traveltime     1
studytime      1
failures       0
famrel         1
freetime       1
goout          1
Dalc           1
Walc           1
health         1
absences       0
G1             0
G2             0
G3             0
dtype: int64

In [ ]:
df_numerical.max()#максимальные значения в каждой числовой колонке датасета

age           22
Medu           4
Fedu           4
traveltime     4
studytime      4
failures       3
famrel         5
freetime       5
goout          5
Dalc           5
Walc           5
health         5
absences      32
G1            19
G2            19
G3            19
dtype: int64

Курсив

1.   **Ответ 1.** mean	16.744222	2.514638	2.306626	1.568567	1.930663	0.221880	3.930663	3.180277	3.184900	1.502311	2.280431	3.536210	3.659476	11.399076	11.570108	11.906009
2.   **Ответ 2.1** min	15.000000	0.000000	0.000000	1.000000	1.000000	0.000000	1.000000	1.000000	1.000000	1.000000	1.000000	1.000000	0.000000	0.000000	0.000000	0.000000

  **Ответ 2.2** max	22.000000	4.000000	4.000000	4.000000	4.000000	3.000000	5.000000	5.000000	5.000000	5.000000	5.000000	5.000000	32.000000	19.000000	19.000000	19.000000



##Типизация
Присвойте соответствующим колонкам один из типов (category, float, int)

In [ ]:
df.school = df.school.astype("category")#Место для вашего кода
df.sex = df.sex.astype("category")
df.address = df.address.astype("category")
df.famsize = df.famsize.astype("category")
df.Pstatus = df.Pstatus.astype("category")
df.guardian = df.guardian.astype("category")
df.schoolsup = df.schoolsup.astype("category")
df.famsup = df.famsup.astype("category")
df.paid = df.paid.astype("category")
df.activities = df.activities.astype("category")
df.nursery = df.nursery.astype("category")
df.higher = df.higher.astype("category")
df.internet = df.internet.astype("category")
df.romantic = df.romantic.astype("category")
df.Mjob = df.Mjob.astype("category")
df.Fjob = df.Fjob.astype("category")
df.reason = df.reason.astype("category")
df.age = df.age.astype("int64")
df.Medu = df.Medu.astype("int64")
df.Fedu = df.Fedu.astype("int64")
df.traveltime = df.traveltime.astype("int64")
df.studytime = df.studytime.astype("int64")
df.failures = df.failures.astype("int64")
df.famrel = df.famrel.astype("int64")
df.freetime = df.freetime.astype("int64")
df.goout = df.goout.astype("int64")
df.Dalc = df.Dalc.astype("int64")
df.Walc = df.Walc.astype("int64")
df.health = df.health.astype("int64")
df.absences = df.absences.astype("int64")
df.G1 = df.G1.astype("int64")
df.G2 = df.G2.astype("int64")
df.G3 = df.G3.astype("int64")
df.info()#Вывести типы колонок тут

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 649 entries, 0 to 648
Data columns (total 33 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   school      649 non-null    category
 1   sex         649 non-null    category
 2   age         649 non-null    int64   
 3   address     649 non-null    category
 4   famsize     649 non-null    category
 5   Pstatus     649 non-null    category
 6   Medu        649 non-null    int64   
 7   Fedu        649 non-null    int64   
 8   Mjob        649 non-null    category
 9   Fjob        649 non-null    category
 10  reason      649 non-null    category
 11  guardian    649 non-null    category
 12  traveltime  649 non-null    int64   
 13  studytime   649 non-null    int64   
 14  failures    649 non-null    int64   
 15  schoolsup   649 non-null    category
 16  famsup      649 non-null    category
 17  paid        649 non-null    category
 18  activities  649 non-null    category
 19  nursery 

Какие признаки вы можете выделить в своем датасете (категориальные, числовые и др., **минимум 3 разных**, помните что можно также выделить признаки более точно, например числовые - порядковые, количественные и т.д.)?

**ОТВЕТ:**
1. Бинарные: sex, schoolsup,	famsup,	paid, activities,	nursery, higher,	internet,	romantic, school, address, famsize, Pstatus
2. Категориальные: guardian, Mjob,	Fjob,	reason,	guardian
4. Числовые: age, Medu, Fedu, traveltime, studytime, failures, famrel,	freetime,	goout,	Dalc,	Walc,	health,	absences,	G1,	G2,	G3

##Фильтрация
Офильтруйте 3 колонки вашего датасета и выведите результат

In [ ]:
f1=df[df['sex'] == 'F']#Фильтр колонки 1
f1

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   at_home   teacher   
1       GP   F   17       U     GT3       T     1     1   at_home     other   
2       GP   F   15       U     LE3       T     1     1   at_home     other   
3       GP   F   15       U     GT3       T     4     2    health  services   
4       GP   F   16       U     GT3       T     3     3     other     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
642     MS   F   17       U     GT3       T     4     3   teacher     other   
643     MS   F   18       R     GT3       T     4     4   teacher   at_home   
644     MS   F   19       R     GT3       T     2     3  services     other   
645     MS   F   18       U     LE3       T     3     1   teacher  services   
646     MS   F   18       U     GT3       T     1     1     other     other   

         reason guardian  traveltime  studytime  failures schoolsup famsup  \
0        course   mother           2          2         0       yes     no   
1        course   father           1          2         0        no    yes   
2         other   mother           1          2         0       yes     no   
3          home   mother           1          3         0        no    yes   
4          home   father           1          2         0        no    yes   
..          ...      ...         ...        ...       ...       ...    ...   
642       other   mother           2          2         0        no     no   
643  reputation   mother           3          1         0        no    yes   
644      course   mother           1          3         1        no     no   
645      course   mother           1          2         0        no    yes   
646      course   mother           2          2         0        no     no   

    paid activities nursery higher internet romantic  famrel  freetime  goout  \
0     no         no     yes    yes       no       no       4         3      4   
1     no         no      no    yes      yes       no       5         3      3   
2     no         no     yes    yes      yes       no       4         3      2   
3     no        yes     yes    yes      yes      yes       3         2      2   
4     no         no     yes    yes       no       no       4         3      2   
..   ...        ...     ...    ...      ...      ...     ...       ...    ...   
642   no         no     yes    yes      yes       no       5         5      4   
643   no        yes     yes    yes      yes      yes       4         4      3   
644   no        yes      no    yes      yes       no       5         4      2   
645   no         no     yes    yes      yes       no       4         3      4   
646   no        yes     yes    yes       no       no       1         1      1   

     Dalc  Walc  health  absences  G1  G2  G3  
0       1     1       3         4   0  11  11  
1       1     1       3         2   9  11  11  
2       2     3       3         6  12  13  12  
3       1     1       5         0  14  14  14  
4       1     2       5         0  11  13  13  
..    ...   ...     ...       ...  ..  ..  ..  
642     1     1       1         0   6   9  11  
643     2     2       5         4   7   9  10  
644     1     2       5         4  10  11  10  
645     1     1       1         4  15  15  16  
646     1     1       5         6  11  12   9  

[383 rows x 33 columns]

In [ ]:
f2=df[df['age'] >18]#Фильтр колонки 2
f2

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
127     GP   F   19       U     GT3       T     0     1   at_home     other   
166     GP   M   19       U     GT3       T     3     2  services   at_home   
224     GP   F   19       U     GT3       T     3     3     other     other   
279     GP   M   22       U     GT3       T     3     1  services  services   
290     GP   F   19       U     GT3       T     2     2  services  services   
295     GP   M   19       U     LE3       A     4     3  services   at_home   
299     GP   F   20       R     GT3       T     2     1     other     other   
311     GP   F   19       U     GT3       T     3     3     other  services   
322     GP   F   19       R     GT3       T     3     2  services  services   
341     GP   F   19       U     GT3       T     4     4    health     other   
350     GP   M   19       R     LE3       T     2     1   at_home  services   
351     GP   M   20       U     GT3       A     3     2  services     other   
352     GP   M   19       R     GT3       T     3     3     other  services   
361     GP   M   19       U     GT3       T     4     2    health     other   
369     GP   F   19       U     GT3       T     1     1     other     other   
370     GP   F   19       U     LE3       A     1     1     other     other   
384     GP   M   19       R     GT3       T     3     2   at_home  services   
405     GP   F   19       U     LE3       A     2     3   at_home     other   
407     GP   F   21       U     LE3       T     4     4     other     other   
413     GP   M   21       R     LE3       T     1     1   at_home     other   
414     GP   F   20       R     GT3       T     1     1     other     other   
415     GP   F   19       U     GT3       T     4     4   teacher     other   
419     GP   M   19       U     GT3       T     2     1     other     other   
421     GP   F   20       U     GT3       T     1     0     other     other   
480     MS   F   19       U     GT3       T     2     3   at_home  services   
488     MS   F   19       U     GT3       T     1     1   at_home  services   
491     MS   F   19       U     GT3       T     1     1     other     other   
504     MS   F   19       U     LE3       T     2     2     other     other   
506     MS   F   20       U     GT3       T     3     3   at_home  services   
552     MS   M   19       U     GT3       T     1     1     other     other   
568     MS   M   19       U     GT3       T     3     2   at_home  services   
571     MS   M   19       U     GT3       T     2     1   at_home     other   
574     MS   M   20       R     GT3       T     1     1     other     other   
577     MS   M   19       R     GT3       T     1     1     other  services   
580     MS   M   19       R     GT3       T     1     1     other     other   
582     MS   F   19       U     LE3       A     1     1   at_home     other   
605     MS   F   19       U     GT3       T     1     1   at_home  services   
610     MS   F   19       R     GT3       A     1     1   at_home   at_home   
632     MS   F   19       R     GT3       T     1     1   at_home     other   
639     MS   M   19       R     GT3       T     1     1     other  services   
644     MS   F   19       R     GT3       T     2     3  services     other   

         reason guardian  traveltime  studytime  failures schoolsup famsup  \
127      course    other           1          2         2        no    yes   
166        home   mother           1          1         0        no    yes   
224  reputation    other           1          4         0        no    yes   
279       other   mother           1          1         3        no     no   
290        home   mother           1          2         0        no    yes   
295  reputation   mother           1          2         0        no    yes   
299      course    other           2          2         0        no    yes   
311        home    other           1          2         0    

In [ ]:

f3=df[df.reason.str.startswith('h')]#Фильтр колонки 3
f3

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
3       GP   F   15       U     GT3       T     4     2    health  services   
4       GP   F   16       U     GT3       T     3     3     other     other   
6       GP   M   16       U     LE3       T     2     2     other     other   
7       GP   F   17       U     GT3       A     4     4     other   teacher   
8       GP   M   15       U     LE3       A     3     2  services     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
631     MS   F   18       R     GT3       T     1     1     other     other   
634     MS   F   18       U     GT3       T     3     3     other     other   
636     MS   M   18       U     GT3       T     4     4   teacher   teacher   
638     MS   M   17       U     GT3       T     2     3     other  services   
640     MS   M   18       R     GT3       T     4     2     other     other   

    reason guardian  traveltime  studytime  failures schoolsup famsup paid  \
3     home   mother           1          3         0        no    yes   no   
4     home   father           1          2         0        no    yes   no   
6     home   mother           1          2         0        no     no   no   
7     home   mother           2          2         0       yes    yes   no   
8     home   mother           1          2         0        no    yes   no   
..     ...      ...         ...        ...       ...       ...    ...  ...   
631   home   mother           4          3         0        no     no   no   
634   home   mother           1          2         0        no     no   no   
636   home   father           1          2         0        no     no   no   
638   home   father           2          2         0        no     no   no   
640   home   father           2          1         1        no     no  yes   

    activities nursery higher internet romantic  famrel  freetime  goout  \
3          yes     yes    yes      yes      yes       3         2      2   
4           no     yes    yes       no       no       4         3      2   
6           no     yes    yes      yes       no       4         4      4   
7           no     yes    yes       no       no       4         1      4   
8           no     yes    yes      yes       no       4         2      2   
..         ...     ...    ...      ...      ...     ...       ...    ...   
631         no     yes    yes      yes       no       4         3      2   
634         no     yes    yes      yes      yes       4         1      3   
636        yes      no    yes      yes       no       3         2      4   
638        yes     yes    yes      yes       no       4         4      3   
640         no     yes    yes       no       no       5         4      3   

     Dalc  Walc  health  absences  G1  G2  G3  
3       1     1       5         0  14  14  14  
4       1     2       5         0  11  13  13  
6       1     1       3         0  13  12  13  
7       1     1       1         2  10  13  13  
8       1     1       1         0  15  16  17  
..    ...   ...     ...       ...  ..  ..  ..  
631     1     2       4         4  10  11  12  
634     1     2       1         1  16  16  16  
636     1     4       2         4  17  18  19  
638     1     1       3         4  14  15  16  
640     4     3       3         0   7   7   0  

[149 rows x 33 columns]

##Группировка
Сгруппируйте датасет по какому-либо категориальному признаку и выведите среднее значение численных колонок (см. метод mean в pandas)

In [ ]:
f4 = df.groupby(by=['sex']).mean()
f4 #Группировка

<ipython-input-16-7c448cec05e1>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  f4 = df.groupby(by=['sex']).mean()


age      Medu      Fedu  traveltime  studytime  failures    famrel  \
sex                                                                             
F    16.788512  2.402089  2.229765    1.543081   2.073107  0.185379  3.864230   
M    16.680451  2.676692  2.417293    1.605263   1.725564  0.274436  4.026316   

     freetime     goout      Dalc      Walc    health  absences         G1  \
sex                                                                          
F    3.052219  3.127937  1.284595  1.937337  3.368146  3.577023  11.637076   
M    3.364662  3.266917  1.815789  2.774436  3.778195  3.778195  11.056391   

            G2         G3  
sex                        
F    11.822454  12.253264  
M    11.206767  11.406015

##Задание для гениев *
Агрегируйте числовой признак по категориальному признаку с использованием метода **без использования** библиотеки pandas, и вычислите медиану и среднее значение для каждой категории.

In [ ]:
#Агрегация